Cython/Build/Cache.py

```diff
@@ -0,0 +1,189 @@
+import sys
+import os
+import hashlib
+import shutil
+import subprocess
+from ..Utils import safe_makedirs, cached_function
+import zipfile
+from .. import __version__
+
+try:
+    import zlib
+    zipfile_compression_mode = zipfile.ZIP_DEFLATED
+except ImportError:
+    zipfile_compression_mode = zipfile.ZIP_STORED
+
+try:
+    import gzip
+    gzip_open = gzip.open
+    gzip_ext = '.gz'
+except ImportError:
+    gzip_open = open
+    gzip_ext = ''
+
+
+join_path = cached_function(os.path.join)
+
+@cached_function
+def file_hash(filename):
+    path = os.path.normpath(filename)
+    prefix = ('%d:%s' % (len(path), path)).encode("UTF-8")
+    m = hashlib.sha1(prefix)
+    with open(path, 'rb') as f:
+        data = f.read(65000)
+        while data:
+            m.update(data)
+            data = f.read(65000)
+    return m.hexdigest()
+
+
+@cached_function
+def get_cython_cache_dir():
+    r"""
+    Return the base directory containing Cython's caches.
+
+    Priority:
+
+    1. CYTHON_CACHE_DIR
+    2. (OS X): ~/Library/Caches/Cython
+       (posix not OS X): XDG_CACHE_HOME/cython if XDG_CACHE_HOME defined
+    3. ~/.cython
+
+    """
+    if 'CYTHON_CACHE_DIR' in os.environ:
+        return os.environ['CYTHON_CACHE_DIR']
+
+    parent = None
+    if os.name == 'posix':
+        if sys.platform == 'darwin':
+            parent = os.path.expanduser('~/Library/Caches')
+        else:
+            # this could fallback on ~/.cache
+            parent = os.environ.get('XDG_CACHE_HOME')
+
+    if parent and os.path.isdir(parent):
+        return os.path.join(parent, 'cython')
+
+    # last fallback: ~/.cython
+    return os.path.expanduser(os.path.join('~', '.cython'))
+
+
+class Cache:
+
+    def __init__(self, cache, cache_size=None):
+        if cache is True:
+            self.enabled = True
+            self.path = os.path.join(get_cython_cache_dir(), 'compiler')
+        elif cache:
+            self.path = cache
+            self.enabled = True
+        else:
+            self.path = None
+            self.enabled = False
+        self.cache_size = cache_size if cache_size is not None else 1024 * 1024 * 100
+        if self.enabled:
+            if not os.path.exists(self.path):
+                os.makedirs(self.path)
+
+    def transitive_fingerprint(self, filename, dependencies, compilation_options, language, py_limited_api=False, np_pythran=False):
```

Comment:
This function (it's not even a method) now receives lots of arguments and does very little with them. That suggests that it receives too many arguments.

For example, it receive the complete set of compilation options, only to call `.get_fingerprint()` on them. Aren't they static at the time the cache is created? Do we need to take a new fingerprint of them for each file? Or is it something that a staticmethod factory function (`Cache.from_options(compile_options)`) could do at init time?

Similarly, it receives the dependencies instance, just to call `.all_dependencies()` on it with the filename that it also receives from outside. Maybe it only needs to know the list of dependencies, and not how to determine them? The latter part seems to belong into `Dependencies.py`.

Finally, it also receives three separate options (`language`, `py_limited_api`, `np_pythran`), just to turn them into a tuple and stringify them. I think `language` is really a flag as well: C++ or not. So these are all flags that could be passed as a single parameter, maybe as a flags dataclass with defaults. The fact that the fingerprinting then uses `c++` or `c` (or often `None`, but we should normalise that) then is an implementation detail local to this function again.

Reply:
> For example, it receive the complete set of compilation options, only to call .get_fingerprint() on them. Aren't they static at the time the cache is created? Do we need to take a new fingerprint of them for each file? Or is it something that a staticmethod factory function (Cache.from_options(compile_options)) could do at init time?

I am not sure if it can be done at init time. `options` are changed for every file compiled - see
https://github.com/cython/cython/blob/9ba08c73193fd850ea4465e1928bdc40c60dae0e/Cython/Build/Dependencies.py#L996-L1002

> Similarly, it receives the dependencies instance, just to call .all_dependencies() on it with the filename that it also receives from outside. Maybe it only needs to know the list of dependencies, and not how to determine them? The latter part seems to belong into Dependencies.py.

Done

> Finally, it also receives three separate options (language, py_limited_api, np_pythran), just to turn them into a tuple and stringify them. I think language is really a flag as well: C++ or not. So these are all flags that could be passed as a single parameter, maybe as a flags dataclass with defaults. The fact that the fingerprinting then uses c++ or c (or often None, but we should normalise that) then is an implementation detail local to this function again.

Done